packages/table/src/common/batcher.ts

```diff
@@ -0,0 +1,182 @@
+/**
+ * Copyright 2017 Palantir Technologies, Inc. All rights reserved.
+ * Licensed under the BSD-3 License as modified (the “License”); you may obtain a copy
+ * of the license at https://github.com/palantir/blueprint/blob/master/LICENSE
+ * and https://github.com/palantir/blueprint/blob/master/PATENTS
+ */
+
+import { requestIdleCallback } from "./requestIdleCallback";
+
+export interface IIndex<T> {
+    [key: string]: T;
+}
+
+export type SimpleStringifyable = string | number | null;
+
+export type Callback = () => void;
+
+/**
+ * This class helps batch updates to large lists.
+ *
+ * For example, if your react component has many children, updating them all at
+ * once may cause jank when reconciling the DOM. This class helps you update
+ * only a few per frame.
+ *
+ * A typical usage would be:
+ *
+ * ```typescript
+ *
+ * public renderChildren = (allChildrenKeys: string[]) => {
+ *
+ *     batcher.startNewBatch();
+ *
+ *     allChildrenKeys.forEach((prop1, index) => {
+ *         batcher.addArgsToBatch(prop1, "prop2", index);
+ *     });
+ *
+ *     batcher.removeOldAddNew((prop1: string, prop2: string, other: number) => {
+ *         return <Child prop1={prop1} prop2={prop2} other={other} />;
+ *     });
+ *
+ *     if (!batcher.isDone()) {
+ *         batcher.idleCallback(this.forceUpdate());
+ *     }
+ *
+ *     const currentChildren = batcher.getList();
+ *     return currentChildren;
+ * }
+ *
+ * ```
+ */
+export class Batcher<T> {
+    public static DEFAULT_ADD_LIMIT = 20;
+    public static DEFAULT_REMOVE_LIMIT = 20;
+
+    private currentObjects: IIndex<T> = {};
+    private batchArgs: IIndex<any[]> = {};
+    private done = true;
+    private callback: Callback;
+
+    /**
+     * Resets the "batch" and "current" sets. This essentially clears the cache
+     * and prevents accidental re-use of "current" objects.
+     */
+    public reset() {
+        this.batchArgs = {};
+        this.currentObjects = {};
+    }
+
+    /**
+     * Starts a new "batch" argument set
+     */
+    public startNewBatch() {
+        this.batchArgs = {};
+    }
+
+    /**
+     * Stores the variadic arguments to be later batched together.
+     *
+     * The arguments must be simple stringifyable objects.
+     */
+    public addArgsToBatch(...args: SimpleStringifyable[]) {
+        this.batchArgs[args.join("-")] = args;
+    }
+
+    /**
+     * Compares the set of "batch" arguments to the "current" set. Creates any
+     * new objects using the callback as a factory. Removes old objects.
+     *
+     * Arguments are in the "current" set but were not part of the last "batch"
+     * set are considered candidates for removal. Similarly, Arguments that are
+     * part of the "batch" set but not the "current" set are candidates for
+     * addition.
+     *
+     * The number of objects added and removed maybe limited with the `..Limit`
```

Comment:
ok comments in this file need some general editing, i'm not going to call out each typo here :book: